### PR TITLE
Transaction Application

### PR DESCRIPTION
**Pull request by Gambit142**
## Branch Implementation 💻 

I created a Ruby on Rails application that allows the user to:
- register and log in, so that the data is private to them.
- introduce new transactions associated with a category.
- see the money spent on each category.

#### Design
- I followed these [design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), including:
  - Colors.
  - Typography: font size and weight.
  - Layout: composition and space between elements.

The original design idea was from [Gregoire Vella on Behance](https://www.behance.net/gregoirevella) and I acknowledged her in my `README.md` file

#### Interactions
- **_Splash screen_**
  - The splash screen is a simple page with the name of my app and links to the `signup` and  `login` pages.

- **_Sign up and login pages_**
  - The user can register in the app with full name, email, and password (all mandatory).
  - The user can log into the app using email and password.
  - If the user is not logged in, they can't access pages that require the user to be logged in.

- **_Home page (categories page)_**
  - When the user logs in, they are presented with the categories page.
  - For each category, the user can see their name, icon, and the total amount of all the transactions that belong to that category.
  - When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
  - There is a button `" add a new category "` at the bottom that brings the user to the page to create a new category.

- **_Transactions page_**
  - For a given category, the list of transactions is presented and ordered by the most recent.
  - At the top of the page, the user can see the total amount for the category `(sum of all of the amounts of the transactions in that category)`.
  - There is a button `" add a new transaction "` at the bottom that brings the user to the page to create a new transaction.
  - When the user clicks on the `"Back"` button `(<)`, the user navigates to the home page.

- **_" Add a new category" page_**
  - The user can fill out a form to create a new category, indicating their name and icon (both mandatory).
  - The user can click (or tap) the `"Save"` button to create the new category and is taken to the home page on success.
  - When the user clicks on the `"Back"` button `(<)`, the user navigates to the home page.

- **_"Add a new transaction"_** page
  - The user can fill out a form to create a new transaction with:
    - name (mandatory)
    - amount (mandatory)
    - categories (mandatory at least one)
  - The user can click (or tap) the `"Save"` button to create the new category and is taken to the transactions page for that category.
  - When the user clicks on the `"Back"` button `(<)`, the user navigates to the transactions page for that category.

#### Testing requirements
- I created unit and integration tests for the most important components of the application.

#### Technical requirements

- I used `Postgres` as your database.
- I used `devise` for authentication.
- I used `cancancan` for authorization.
- I validated all user input to make sure that anyone with bad intentions cannot compromise the app.
- I used `.erb` view template engine.
- I deployed the project using `Heroku`

### Video Presentation
- The Video documentation for this project can be found [HERE](https://www.loom.com/share/3678f552a51743139a2df6cdea37b2de)

### Project Requirements
- The project requirements for this capstone project can be found [HERE](https://github.com/microverseinc/curriculum-rails/blob/main/capstone/rails_capstone.md)

### CODE REVIEW

File: app/assets/stylesheets/category.scss [hunk 1]
```diff
@@ -0,0 +1,122 @@
+$italianno: 'Italianno', cursive;
+$italiana: 'Italiana', serif;
+$roboto: 'Roboto', sans-serif;
+$blue: #3778c2;
```
Comment (Dipeshtwis) [inline app/assets/stylesheets/category.scss:4]: - [ ] I would recommend using the same font which is recommended in the[ design](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=).

File: app/assets/stylesheets/application.css [hunk 2]
```diff
@@ -13,3 +13,96 @@
...
+  color: white;
+}
+
+.devise-btn {
+  margin: 0 2rem;
+  width: 20rem;
+  text-align: center;
+  padding: 1rem;
+  font-family: 'Italiana', serif;
+  font-size: 1.5rem;
```
Comment (Dipeshtwis) [inline app/assets/stylesheets/application.css:78]: - [ ] Here as well. I would recommend using the same font which is recommended in the[ design](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=).

**Review by Dipeshtwis — CHANGES_REQUESTED** [review]
Hi @Gambit142, 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: App is deployed on Heroku.
:+1: All tests are passing. :heavy_check_mark: 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


 ## Required Changes ♻️

**_Check the comments under the review_**

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

- N.A.

Cheers and Happy coding!👏👏👏
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**_Dipesh Kumar_**


------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project(check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/records_controller.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    end
+  end
+
+  def current_category
+    @category = current_user.categories.includes(:records).find(params[:category_id])
+  end
+
+  def record_params
+    params.require(:add_record).permit(:name, :amount, category_ids: [])
+  end
```
[the remark this answers is not in the document]
Comment (Gambit142) [reply app/controllers/records_controller.rb]: Thank you for the feedback, I have implemented this change 👍

File: app/assets/stylesheets/category.scss (hunk 1 above)
Comment (Gambit142) [reply app/assets/stylesheets/category.scss]: Thank you for the feedback, I have implemented this change 👍

File: app/assets/stylesheets/application.css (hunk 2 above)
Comment (Gambit142) [reply app/assets/stylesheets/application.css]: Thank you for the feedback, I have implemented this change 👍

**Review by bimbolabuari — APPROVED** [review]
# Project Approved 🚀

Hello @Gambit142🥇,

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**